File: front/react/src/util/shadowing/currentSentence.test.ts
import { describe, expect, it } from 'vitest'
import type { Sentence } from '@/types/api'
import { getCurrentSentenceIndex } from './currentSentence'

const sentences: Sentence[] = [
  { index: 0, tokens: [], chinese: 'one', start_ms: 1000, end_ms: 2000 },
  { index: 1, tokens: [], chinese: 'two', start_ms: 3000, end_ms: 4000 },
  { index: 2, tokens: [], chinese: 'three', start_ms: 4000, end_ms: 5000 },
]

describe('getCurrentSentenceIndex', () => {
  it('returns null for empty list', () => {
    expect(getCurrentSentenceIndex([], 1000)).toBeNull()
  })

  it('uses first sentence before first start', () => {
    expect(getCurrentSentenceIndex(sentences, 0)).toBe(0)
  })

  it('keeps previous sentence in gaps', () => {
    expect(getCurrentSentenceIndex(sentences, 2500)).toBe(0)
  })

  it('uses next sentence at exact start boundary', () => {
    expect(getCurrentSentenceIndex(sentences, 3000)).toBe(1)
  })

  it('keeps last sentence after media ends', () => {
    expect(getCurrentSentenceIndex(sentences, 6000)).toBe(2)
  })

  it('skips invalid sentence timings', () => {
    const invalidLeading: Sentence[] = [
      { index: 9, tokens: [], chinese: 'bad', start_ms: 0, end_ms: 0 },
      { index: 10, tokens: [], chinese: 'ok', start_ms: 1000, end_ms: 2000 },
    ]

    expect(getCurrentSentenceIndex(invalidLeading, 500)).toBe(10)
  })
})

File: internal/module/shadowing/current_sentence_test.go
package shadowing

import (
	"testing"

	"japanese-learning-app/internal/module/lesson"
)

func TestCurrentSentenceIndex(t *testing.T) {
	sentences := []lesson.Sentence{
		{Index: 10, StartMS: 1000, EndMS: 2000},
		{Index: 11, StartMS: 3000, EndMS: 4000},
		{Index: 12, StartMS: 4500, EndMS: 5500},
	}

	tests := []struct {
		name          string
		sentences     []lesson.Sentence
		currentTimeMS int64
		wantIndex     int
		wantOK        bool
	}{
		{
			name:          "empty list",
			sentences:     nil,
			currentTimeMS: 1000,
			wantIndex:     -1,
			wantOK:        false,
		},
		{
			name:          "before first sentence returns first index",
			sentences:     sentences,
			currentTimeMS: 500,
			wantIndex:     10,
			wantOK:        true,
		},
		{
			name:          "exact start uses that sentence",
			sentences:     sentences,
			currentTimeMS: 3000,
			wantIndex:     11,
			wantOK:        true,
		},
		{
			name:          "inside sentence returns that sentence",
			sentences:     sentences,
			currentTimeMS: 3500,
			wantIndex:     11,
			wantOK:        true,
		},
		{
			name:          "gap returns previous valid sentence",
			sentences:     sentences,
			currentTimeMS: 2500,
			wantIndex:     10,
			wantOK:        true,
		},
		{
			name:          "after last sentence returns last valid index",
			sentences:     sentences,
			currentTimeMS: 9000,
			wantIndex:     12,
			wantOK:        true,
		},
		{
			name: "invalid sentences are skipped",
			sentences: []lesson.Sentence{
				{Index: 1, StartMS: 0, EndMS: 0},
				{Index: 2, StartMS: 1000, EndMS: 900},
				{Index: 3, StartMS: 2000, EndMS: 3000},
			},
			currentTimeMS: 1500,
			wantIndex:     3,
			wantOK:        true,
		},
		{
			name: "all invalid sentences return not found",
			sentences: []lesson.Sentence{
				{Index: 1, StartMS: 0, EndMS: 0},
				{Index: 2, StartMS: 1000, EndMS: 999},
			},
			currentTimeMS: 1500,
			wantIndex:     -1,
			wantOK:        false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gotIndex, gotOK := CurrentSentenceIndex(tt.sentences, tt.currentTimeMS)
			if gotIndex != tt.wantIndex || gotOK != tt.wantOK {
				t.Fatalf("CurrentSentenceIndex() = (%d, %v), want (%d, %v)", gotIndex, gotOK, tt.wantIndex, tt.wantOK)
			}
		})
	}
}
